packages/service-library/src/servicelib/redis/_project_lock.py

```diff
@@ -0,0 +1,99 @@
+import functools
+from collections.abc import Awaitable, Callable, Coroutine
+from typing import Any, Final, ParamSpec, TypeVar
+
+from models_library.projects import ProjectID
+from models_library.projects_access import Owner
+from models_library.projects_state import ProjectLocked, ProjectStatus
+
+from ._client import RedisClientSDK
+from ._decorators import exclusive
+from ._errors import CouldNotAcquireLockError, ProjectLockError
+
+_PROJECT_REDIS_LOCK_KEY: Final[str] = "project_lock:{}"
+
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+def with_project_locked(
+    redis_client: RedisClientSDK | Callable[..., RedisClientSDK],
+    *,
+    project_uuid: str | ProjectID,
+    status: ProjectStatus,
+    owner: Owner | None,
+    notification_cb: Callable[[], Awaitable[None]] | None,
+) -> Callable[
+    [Callable[P, Coroutine[Any, Any, R]]], Callable[P, Coroutine[Any, Any, R]]
+]:
+    """creates a distributed auto sustained Redis lock for project with project_uuid, keeping its status and owner in the lock data
+
+    Arguments:
+        redis_client -- the client to use to access redis
+        project_uuid -- the project UUID
+        status -- the project status
+        owner -- the owner of the lock (default: {None})
+        notification_cb -- an optional notification callback that will be called AFTER the project is locked and AFTER it was unlocked
+
+    Returns:
+        the decorated function return value
+
+    Raises:
+        raises anything from the decorated function and from the optional notification callback
+    """
+
+    def _decorator(
+        func: Callable[P, Coroutine[Any, Any, R]],
+    ) -> Callable[P, Coroutine[Any, Any, R]]:
+        @functools.wraps(func)
+        async def _wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+            @exclusive(
+                redis_client,
+                lock_key=_PROJECT_REDIS_LOCK_KEY.format(project_uuid),
+                lock_value=ProjectLocked(
+                    value=True,
+                    owner=owner,
+                    status=status,
+                ).model_dump_json(),
+            )
+            async def _exclusive_func(*args, **kwargs) -> R:
+                if notification_cb is not None:
+                    await notification_cb()
```

Comment:
Q: this callback runs before running. should we have a mechanism to call when finished? 
I have a thought about how to generalize this ...